File: Week-7/Exercise7_5.cpp
#include <iostream>
#include <stack>
using namespace std;
 
void findNextGreaterElements(int arr[], int n) {
    int res[n];
    stack<int> s;
 
    for (int i = n - 1; i >= 0; i--) {
        if (!s.empty()) {
            while (!s.empty() && s.top() <= arr[i]) {
                s.pop();
            }
        }
        res[i] = s.empty() ? -1 : s.top();
        s.push(arr[i]);
    }
 
    for (int i = 0; i < n; i++)
        cout << arr[i] << " --> " << res[i] << endl;
}

int main()
{
    int arr[] = { 2, 7, 3, 5, 4, 6, 8 };
    int n = sizeof(arr) / sizeof(arr[0]);

    findNextGreaterElements(arr, n); 
    return 0;
}

// Time Complexity: O(n)

File: Week-7/Exercise7_5.js
let arr = [6, 8, 0, 1, 3]

// for (let i = 0; i < x.length; i++) {
//     let max = -1;
//     for (let j = i + 1; j < x.length; j++) {
//         if (x[j] > x[i]) {
//             max = x[j];
//             break;
//         }
//     }
//     console.log(max);
// }

// time complexity: O(n^2)


function nextLargerElement(arr, n) {
    var s = [];

    // for (var i = 0; i < arr.length; i++) {
    //     while (s.length > 0 && s[s.length - 1]["value"] < arr[i]) {
    //         var d = s.pop();
    //         arr[d["ind"]] = arr[i];
    //     }
    //     s.push({ "value": arr[i], "ind": i });
    // }

    while (s.length > 0) {
        d = s.pop();
        arr[d["ind"]] = -1;
    }
    return arr;

}

console.log(nextLargerElement(arr, 5));